File: typescript/nucleotide-count/nucleotide-count.ts
export default class NucleotideCount {
  static nucleotideCounts(strand: string): { [letter: string]: number } {
    const counts: { [letter: string]: number } = {
      A: 0,
      C: 0,
      G: 0,
      T: 0
    }

    for (const character of strand) {
      const characterCount = counts[character]
      if (characterCount === undefined) {
        throw new Error('Invalid nucleotide in strand')
      }

      counts[character] = characterCount + 1
    }

    return counts
  }
}

File: fsharp/nucleotide-count/NucleotideCount.fs
module NucleotideCount

let nucleotideCounts strand =
    let initial = Map.ofList [('A', 0); ('C', 0); ('G', 0); ('T', 0)]
    let counts = Seq.countBy id strand |> Seq.fold (fun counts (letter, count) -> Map.add letter count counts) initial
    if counts.Count = 4 then Some counts else None
